src/akushon/action/process/interpolator.cpp

```diff
@@ -0,0 +1,162 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <string>
+#include <vector>
+
+#include "akushon/action/process/interpolator.hpp"
+
+#include "akushon/action/model/action.hpp"
+#include "akushon/action/process/joint_process.hpp"
+#include "tachimawari/joint/model/joint.hpp"
+
+namespace akushon
+{
+
+Interpolator::Interpolator(const std::vector<Action> & actions, const Pose & initial_pose)
+: actions(actions), joint_processes({}), current_pose_index(0),
+  pause_time(0), init_pause(false), start_stop_time(0), init_state(true),
+  current_action_index(0)
+{
+  for (const auto & joint : initial_pose.get_joints()) {
+    joint_processes.insert(
+      {joint.get_id(),
+        JointProcess(joint.get_id(), joint.get_position())});
+  }
+
+  if (actions.size() != 0) {
+    state = START_DELAY;
+  } else {
+    state = END;
+  }
+}
+
+void Interpolator::process(int time)
+{
+  switch (state) {
+    case START_DELAY:
+      {
+        if (init_state) {
+          init_state = false;
+          start_stop_time = time;
+        }
+
+        if ((time - start_stop_time) > (actions[current_action_index].get_start_delay() * 1000)) {
+          change_state(PLAYING);
+        }
+
+        break;
+      }
+
+    case PLAYING:
+      {
+        if (check_for_next()) {
+          if (init_pause) {
+            init_pause = false;
+            pause_time = time;
+          }
+
+          if (current_pose_index == actions[current_action_index].get_pose_count()) {
+            change_state(STOP_DELAY);
+            init_pause = true;
+          } else if ((time - pause_time) >  // NOLINT
+            (actions[current_action_index].get_pose(current_pose_index).get_pause() * 1000))
+          {
+            next_pose();
+            init_pause = true;
+          }
+        }
+
+        break;
+      }
+
+    case STOP_DELAY:
+      {
+        if (init_state) {
+          init_state = false;
+          start_stop_time = time;
+        }
+
+        if ((time - start_stop_time) > (actions[current_action_index].get_stop_delay() * 1000)) {
+          current_action_index++;
+
+          if (current_action_index == actions.size()) {
+            change_state(END);
+          } else {
+            change_state(START_DELAY);
+          }
+        }
+
+        break;
+      }
+  }
+
+  for (auto [id, joint] : joint_processes) {
+    joint_processes.at(id).interpolate();
+  }
+}
+
+bool Interpolator::is_finished() const
+{
+  return state == END;
+}
+
+void Interpolator::next_pose()
+{
+  for (const auto & joint :
+    actions[current_action_index].get_pose(current_pose_index).get_joints())
```

Comment:
can we save `actions[current_action_index]` to shorten content of `for` ?
```c++
const auto & current_action = actions[current_action_index];
for (const auto & joint : current_action.get_pose(current_pose_index).get_joints()) {
  // ...
}
```

or better, create a function from it (if it always get called)
```c++
const auto & get_current_action() const {
  return actions[current_action_index];
}

const auto & get_current_pose() const {
  return get_current_action().get_pose(current_pose_index);
}
```

Reply:
done